### ISSUE TITLE
SP heuristic finds incorrect optimum for simple SP

### ISSUE THREAD

**Issue by pgkirsch** [convexopt/gpkit#274]
This example should solve to an optimum at x = 1, y = 1.

```
import gpkit
from gpkit.shortcuts import *

gpkit.enable_signomials() 

x = Var('x')
y = Var('y')

J = 0.01*((x - 1)**2 + (y - 1)**2) + (x*y - 1)**2

sp = SP(J)
print sp
sol = sp.localsolve(x0={x: 1.1, y: 1.1})

print sol.table()

gpkit.disable_signomials()
```

Output: 
```
In [1]: %run Mike_SP2.py
Optional Python units library (Pint) not installed;unit support disabled.
gpkit.GeometricProgram( # minimize
          -0.02*x + -0.02*y + -2*x*y + 0.01*x**2 + 0.01*y**2 + 1 + x**2*y**2,
          [   # subject to
          ],
          substitutions={  },
          solver="mosek")
Beginning signomial solve.
Using solver 'mosek'
Solving for 2 variables.
Solving took 0.159 seconds.
Solving took 5 GP solves.

Cost
 0.624   

Variables
---------
x : 0.789   
y : 0.789   

Constant sensitivities
----------------------
```
Possibly related issue: when we `print sp` it outputs `-0.02*x + -0.02*y + -2*x*y + 0.01*x**2 + 0.01*y**2 + 1 + x**2*y**2`. Notice that the constant term should be `1.02` instead of `1`.

**Comment by whoburg** [comment]
Todo:

* [ ] fix this
* [ ] add a test
* [ ] also test `sol = sp.localsolve()` (no initial guess)

**Comment by pgkirsch** [comment]
Update: The output appears to be different from when this ticket was first opened.

```
[philippekirschen@kirschens-MBP tests]$ ipython issue_274.py 
Optional Python units library (Pint) not installed; unit support disabled.
gpkit.GeometricProgram( # minimize
          -0.02*x + -0.02*y + -2*x*y + 0.01*x**2 + 0.01*y**2 + 1 + x**2*y**2,
          [   # subject to
          ],
          substitutions={  },
          solver="mosek")
Beginning signomial solve.
Using solver 'mosek'
Solving for 2 variables.
Solving took 0.191 seconds.
Solving took 5 GP solves.

Cost
----
 0.1433

Free variables
--------------
x : 0.789 
y : 0.789 

Swept variables
---------------

Constants
---------

Constant and swept variable sensitivities
-----------------------------------------
```

**Comment by pgkirsch** [comment]
Observation:

```
In [28]: gpkit.enable_signomials()

In [29]: x = Var('x')

In [30]: y = Var('y')

In [31]: x-1
Out[31]: gpkit.Signomial(- + x)

In [32]: x-2
Out[32]: gpkit.Signomial(-2 + x)
```

**Comment by pgkirsch** [comment]
Another observation:

```
In [6]: x-0.999
Out[6]: gpkit.Signomial(- + x)

In [7]: x-0.99
Out[7]: gpkit.Signomial(-0.99 + x)
```

**Comment by pgkirsch** [comment]
Sub-issues to address:
- [ ] The `1` instead of `1.02` constant issue
- [ ] The `gpkit.Signomial(- + x)` issue
- [ ] The SP solution issue (if it still remains)

I am open to splitting these into separate issues if anyone has any strong preferences, but for the time being I will keep them here.

**Comment by pgkirsch** [comment]
The first sub-issue seems unrelated enough and significant enough to merit its own ticket, so I have moved it #316.

**Comment by bqpd** [comment]
The second issue is happening the same place as the first, I'll move it to #316 as well.

**Comment by pgkirsch** [comment]
Thanks

**Comment by bqpd** [comment]
Weirdly, the results on this are better without the `0.01*` out front.

**Comment by pgkirsch** [comment]
And they only get better as the coefficient increases. Fore example, with `10000*` in front, the error (cost) is on the order 1e-6.

**Comment by pgkirsch** [comment]
Getting an error from this

```
import gpkit
from gpkit.shortcuts import *
gpkit.enable_signomials()

x = Var('x')
y = Var('y')

J = 0.01*((x - 1)**2 + (y - 1)**2) 

sp = SP(J)
sol = sp.localsolve(x0={x: 1.1, y: 1.1})

print sol.table()
gpkit.disable_signomials()
```

There is a long error traceback `sol = sp.localsolve(x0={x: 1.1, y: 1.1})`, followed by a rather unsatisfying
```
KeyError: y
```

**Comment by pgkirsch** [comment]
Ok, really weird. I don't get that error without the `0.01` coefficient.

**Comment by bqpd** [comment]
Not getting that error in the refactor, though it's still getting the wrong answer.

**Comment by bqpd** [comment]
Hmmm...it gets the correct answer if there's any constant term in the cost. I wonder if the problem is solver limits on driving log(objective) to minus infinity?

**Comment by whoburg** [comment]
Just a hunch, but I think #296 should take priority -- I wonder if solving that will solve this as well.

**Comment by whoburg** [comment]
PS confirmed the 1-x issue is fixed on the refactor branch

```python
In [3]: x = Variable('x')

In [4]: from gpkit import enable_signomials

In [5]: enable_signomials()

In [6]: x - 1
Out[6]: gpkit.Signomial(-1 + x)
```

**Comment by pgkirsch** [comment]
I will focus on #296 for the time being then.

**Comment by pgkirsch** [comment]
Unfortunately, commit 35d60b2e635850d6dadadeea984aab7b349b9b33 did not fix this issue too.

**Comment by whoburg** [comment]
I guess I should have realized it would not -- this fails for both cvxopt and mosek.

**Comment by whoburg** [comment]
Here's something interesting. Slightly modified version of the original example, with the objective moved to a constraint:

```python
from gpkit import Model, Variable
​
gpkit.enable_signomials() 
x = Variable('x')
y = Variable('y')
z = Variable('z')
​
J = 1*((x - 1)**2 + (y - 1)**2) + (x*y - 1)**2
​
sp = Model(z, [z >= J])
print sp
sol = sp.localsolve()
​
print sol.table()
​
gpkit.disable_signomials()
```
output:
```python
# minimize
    z,
[   # subject to
    z >= -2*x + -2*x*y + -2*y + 3 + x**2 + x**2*y**2 + y**2,
],
    substitutions={  }
Beginning signomial solve.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-e83cd81ff128> in <module>()
     10 sp = Model(z, [z >= J])
     11 print sp
---> 12 sol = sp.localsolve()
     13 
     14 print sol.table()

gpkit/gpkit/model.py in localsolve(self, solver, verbosity, skipfailures, *args, **kwargs)
    279             return self._solve("sp", solver, verbosity, skipfailures, *args, **kwargs)
    280         except ValueError:
--> 281             raise ValueError("'localsolve()' can only be called on models that"
    282                              " contain Signomials, because such"
    283                              " models have only local solutions. Models"

ValueError: 'localsolve()' can only be called on models that contain Signomials, because such models have only local solutions. Models without Signomials have global solutions, so try using 'solve()'.
```

**Comment by bqpd** [comment]
Huh, it doesn't for me. It works, actually...

**Comment by bqpd** [comment]
```python
import gpkit
from gpkit import Model, Variable

gpkit.enable_signomials() 
x = Variable('x')
y = Variable('y')
z = Variable('z')

J = 1*((x - 1)**2 + (y - 1)**2) + (x*y - 1)**2

sp = Model(z, [z >= J])
print sp
sol = sp.localsolve("cvxopt")

print sol.table()

gpkit.disable_signomials()```

**Comment by pgkirsch** [comment]
Yah works for me too

**Comment by whoburg** [comment]
Well that is certainly strange. Here's what's creating the ValueError on my machine (I edited the source code to remove the try/catch in `localsolve` and raise the error directly):

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-75cb0e5eb14d> in <module>()
     11 sp = Model(z, [z >= J])
     12 print sp
---> 13 sol = sp.localsolve("cvxopt")
     14 
     15 print sol.table()

gpkit/gpkit/model.py in localsolve(self, solver, verbosity, skipfailures, *args, **kwargs)
    276         RuntimeWarning if an error occurs in solving or parsing the solution.
    277         """
--> 278         return self._solve("sp", solver, verbosity, skipfailures, *args, **kwargs)
    279             #raise ValueError("'localsolve()' can only be called on models that"
    280             #                  " contain Signomials, because such"

gpkit/gpkit/model.py in _solve(self, programType, solver, verbosity, skipfailures, *args, **kwargs)
    375                 result = self.program.solve(*args, **kwargs)
    376             elif programType == "sp":
--> 377                 result = self.program.localsolve(*args, **kwargs)
    378             sol = self.parse_result(result, constants, mmaps)
    379             solution.append(sol)

gpkit/gpkit/signomial_program.py in localsolve(self, solver, verbosity, x0, reltol, iteration_limit, *args, **kwargs)
    113                and (not (cost and prevcost)
    114                     or abs(prevcost-cost)/(prevcost + cost) > reltol)):
--> 115             gp = self.step(x0, verbosity)
    116             self.gps.append(gp)
    117 

gpkit/gpkit/signomial_program.py in step(self, x0, verbosity)
    165 
    166         gp = GeometricProgram(posy_approxs[0], posy_approxs[1:],
--> 167                               verbosity=verbosity-1)
    168         return gp
    169 

gpkit/gpkit/geometric_program.py in __init__(self, cost, constraints, verbosity)
     44         self.cs = np.hstack((mag(p.cs) for p in self.posynomials))
     45         if not all(self.cs > 0):
---> 46             raise ValueError("GeometricPrograms cannot contain Signomials.")
     47         self.exps = functools_reduce(add, (p.exps for p in self.posynomials))
     48         self.varlocs, self.varkeys = locate_vars(self.exps)

ValueError: GeometricPrograms cannot contain Signomials.
```

**Comment by bqpd** [comment]
Hm, looks like a negative c snuck through? Does the code I posted work for you? (you can change the solver, works with all of 'em here)

**Comment by whoburg** [comment]
Just tried your code above using the latest on master; error is slightly different but still a problem:

```python
python issue274.py
# minimize
    z,
[   # subject to
    z >= -2*x + -2*x*y + -2*y + 3 + x**2 + x**2*y**2 + y**2,
],
    substitutions={  }
Beginning signomial solve.
.../gpkit/gpkit/geometric_program.py:143: RuntimeWarning: overflow encountered in exp
  np.exp(solver_out['primal']).ravel()))
.../gpkit/gpkit/small_scripts.py:38: RuntimeWarning: invalid value encountered in double_scalars
  e = mag(x0[vk]*units * p.diff(vk).sub(x0, require_positive=False).c / p0)
.../gpkit/gpkit/geometric_program.py:51: RuntimeWarning: invalid value encountered in greater
  if not all(self.cs > 0):
Traceback (most recent call last):
  File "issue274.py", line 13, in <module>
    sol = sp.localsolve("cvxopt")
  File ".../gpkit/gpkit/model.py", line 286, in localsolve
    raise ValueError("'localsolve()' can only be called on models that"
ValueError: 'localsolve()' can only be called on models that contain Signomials, because such models have only local solutions. Models without Signomials have global solutions, so try using 'solve()'.
```

**Comment by pgkirsch** [comment]
@whoburg are you only getting that error with cvxopt?

**Comment by whoburg** [comment]
negative, I get it for both cvxopt and mosek. Looks like it's raising early in `localsolve` before the solver even gets run.

**Comment by whoburg** [comment]
ps I've removed `.pyc` files, so it's not that.

**Comment by whoburg** [comment]
shooting from the hip here, but could be interesting to add this as a unit test on a branch, then we'll get three more machines (on jenkins) trying it out

**Comment by bqpd** [comment]
Sounds good.

**Comment by pgkirsch** [comment]
Goddamnit @bqpd

**Comment by pgkirsch** [comment]
This one is at least partly my fault for not saying I was working on it.

**Comment by pgkirsch** [comment]
Interesting, passed mosek on one machine failed on all other machine/solver combinations.

**Comment by bqpd** [comment]
Super wacky.

**Comment by whoburg** [comment]
Agreed re: not making a test; #329  is just debugging

**Comment by pgkirsch** [comment]
This might be a separate issue, but I thought this was interesting

```
import gpkit
from gpkit import Model, Variable
with gpkit.EnableSignomials():
    x = Variable('x')
    z = Variable('z')
    J = 0.01*(x - 1)**2
    sp = Model(z, [z >= J])
    sol = sp.localsolve(iteration_limit=50)


#output

Beginning signomial solve.
{x: 0.099503719020998152, z: 0.0094544313448217156}
{x: 0.30858339675657204, z: 0.0070641292715984388}
{x: 0.55138385186126571, z: 0.004137253654887871}
{x: 0.7558555212372341, z: 0.0020329156059758929}
{x: 0.88772160560639424, z: 0.0008742012957197896}
{x: 0.95412334675043098, z: 0.00034540249534560776}
{x: 0.98237655634014498, z: 0.00013081534457576836}
{x: 0.99340766018349569, z: 4.86639535899045e-05}
{x: 0.99755961776297863, z: 1.7977255827085814e-05}
{x: 0.99910015431588806, z: 6.6236742434287643e-06}
{x: 0.99966868273464993, z: 2.4381000785678302e-06}
{x: 0.99987807689494357, z: 8.9711476347932211e-07}
{x: 0.99995514181077461, z: 3.3005551447630321e-07}
{x: 0.9999834968924961, z: 1.2142408125806915e-07}
{x: 0.99999392875103243, z: 4.4669889151738637e-08}
{x: 0.99999776649946515, z: 1.6433196923939157e-08}
{x: 0.99999917833933127, z: 6.0454438375448823e-09}
{x: 0.9999996977276967, z: 2.2239956551815462e-09}
{x: 0.99999988880020207, z: 8.1816243629095626e-10}
{x: 0.99999995909187633, z: 3.0098516157414271e-10}
{x: 0.9999999849507416, z: 1.1072625650178e-10}
{x: 0.99999999446368726, z: 4.0733912983436912e-11}
{x: 1.0, z: 1.0, \fbox{0}: 1.0}
{x: 0.099503719020998152, z: 0.0094544313448217156}
{x: 0.30858339675657204, z: 0.0070641292715984388}
{x: 0.55138385186126571, z: 0.004137253654887871}
{x: 0.7558555212372341, z: 0.0020329156059758929}
{x: 0.88772160560639424, z: 0.0008742012957197896}
{x: 0.95412334675043098, z: 0.00034540249534560776}
{x: 0.98237655634014498, z: 0.00013081534457576836}
{x: 0.99340766018349569, z: 4.86639535899045e-05}
{x: 0.99755961776297863, z: 1.7977255827085814e-05}
{x: 0.99910015431588806, z: 6.6236742434287643e-06}
{x: 0.99966868273464993, z: 2.4381000785678302e-06}
{x: 0.99987807689494357, z: 8.9711476347932211e-07}
{x: 0.99995514181077461, z: 3.3005551447630321e-07}
{x: 0.9999834968924961, z: 1.2142408125806915e-07}
{x: 0.99999392875103243, z: 4.4669889151738637e-08}
{x: 0.99999776649946515, z: 1.6433196923939157e-08}
{x: 0.99999917833933127, z: 6.0454438375448823e-09}
{x: 0.9999996977276967, z: 2.2239956551815462e-09}
{x: 0.99999988880020207, z: 8.1816243629095626e-10}
{x: 0.99999995909187633, z: 3.0098516157414271e-10}
{x: 0.9999999849507416, z: 1.1072625650178e-10}
{x: 0.99999999446368726, z: 4.0733912983436912e-11}
{x: 1.0, z: 1.0, \fbox{1}: 1.0}
{x: 0.099503719020998152, z: 0.0094544313448217156}
{x: 0.30858339675657204, z: 0.0070641292715984388}
{x: 0.55138385186126571, z: 0.004137253654887871}
{x: 0.7558555212372341, z: 0.0020329156059758929}
Solving took 50 GP solves and 0.828 seconds.

Cost
----
 0.002033

Free Variables
--------------
x : 0.7559   
z : 0.002033 

```

Unsurprisingly, if you set `iteration_limit=22`, the answer is correct:

```
Cost
----
 4.073e-11

Free Variables
--------------
x : 1         
z : 4.073e-11 
```

tl;dr the while loop exit condition in `localsolve` doesn't seem to be doing its job.

**Comment by bqpd** [comment]
There's a cycle! Haha, weird...maybe some of those are feasibility solves?
(which aren't permitted to be the last solve)

**Comment by pgkirsch** [comment]
Alas no it doesn't seem like the `except` statement in `localsolve` ever gets activated so they are all plain vanilla `gp.solve`s.

**Comment by bqpd** [comment]
Weirdddd. Is the cost difference really staying above reltol?

**Comment by pgkirsch** [comment]
Is this what you mean?
(the try is necessary for when `prevcost=None`)

```
 try:
       print abs(prevcost-cost)/(prevcost + cost)
except:
       pass
```

in which case yes...

```
# output

0.144704016817
0.261295916399
0.341050295372
0.398578505591
0.433582450504
0.450607123003
0.457720705538
0.460476303345
0.461510257991
0.461893445602
0.462034797781
0.462086850697
0.462106007011
0.462113054959
0.462115648176
0.462116602485
0.462116949706
0.462117078157
0.462117132357
0.462117195347

```

**Comment by pgkirsch** [comment]
By the way please can you explain the first part of that conditional?

```
        while (iterations < iteration_limit
               and (not (cost and prevcost)
                    or abs(prevcost-cost)/(prevcost + cost) > reltol)):
```

(starting from `not`)

**Comment by bqpd** [comment]
Oh, that's catching when those costs are None, such as the first two
iterations and whenever a feasibility gp is used.

Re: costs, hmmmmmm. Perhaps we should make that denominator (prevcost +
cost + 1e-10?)

**Comment by pgkirsch** [comment]
Doesn't seem to help unfortunately

**Comment by bqpd** [comment]
A large enough denominator certainly would :p I think the problem is that prevcost+cost is approaching 0 just as fast as cost is approaching 0

**Comment by pgkirsch** [comment]
Oh wait you were right, I think the end of each "cycle" is a feasible solve...

**Comment by pgkirsch** [comment]
Putting 1e-6 in the denominator does indeed solve the problem.

**Comment by pgkirsch** [comment]
@whoburg @bqpd any suggestions for things I can try, I'm running out of ideas...

**Comment by bqpd** [comment]
The feasible-solve is interesting; what's the gp just before that in model.program? (it should be the infeasible one)

**Comment by pgkirsch** [comment]
Mm no sorry, I meant the original issue, I think you're referring to what is now #336.

**Comment by pgkirsch** [comment]
Although I can certainly look into that more, if you don't think the proposed PR (#337) is satisfactory

**Comment by bqpd** [comment]
Oops, my mistake. I'll try and reproduce this on a Windows machine.

**Comment by whoburg** [comment]
I'm not yet personally convinced that adding `1e-whatever` is a good solution. Note further discussion at #329. We need to first figure out why things behave so differently with J in the objective vs constraints.

**Comment by whoburg** [comment]
I still think that SPs with optimal objective values equal to (or less than) 0 might be fundamentally ill-posed. That's because the monomial approximations become ill-posed...

**Comment by whoburg** [comment]
oops my last two comments should maybe be in #336

**Comment by bqpd** [comment]
I think I figured that out. When forming approximations, it currently assumes that you're approximating (p)/(1+n), where p is the posynomial portion and n is the negynomial. This isn't true for a signomial cost function. The cost function (I believe?) should just be p/n.

**Comment by pgkirsch** [comment]
Inspecting the first GP approximation for both cases (case 1: J in objective, case 2: J in constraints)....

Writing these just as a normal GP to be solved....

```
# case 1
from gpkit.shortcuts import *
x = Var('x')
y = Var('y')
J = 0.00329*x**-0.66*y**1.3 + 0.00329*x**1.3*y**-0.66 + 0.329*x**1.3*y**1.3 + 0.336*x**-0.66*y**-0.66
gp = Model(J)
sol = gp.solve()
```

```
# case 2
from gpkit.shortcuts import *
x = Var('x')
y = Var('y')
z = Var('z')
J = 0.00329*x**-0.66*y**1.3*z**-0.33 + 0.00329*x**1.3*y**-0.66*z**-0.33 + 0.329*x**1.3*y**1.3*z**-0.33 + 0.336*x**-0.66*y**-0.66*z**-0.33
gp = Model(z, [J <=1])
sol = gp.solve()
```

Equivalent GPs, right?
Except the outpuut is different 

```
# case 1
Using solver 'mosek'
Solving for 2 variables.
Solving took 0.0143 seconds.
free variable too sensitive: S_{y} = 6.40e-01
free variable too sensitive: S_{x} = 6.40e-01

Cost
----
 0.6379

Free Variables
--------------
x : 0.8443 
y : 0.8443 

Sensitivities
-------------
x : 0.64 
y : 0.64 
```

```
# case 2
Using solver 'mosek'
Solving for 3 variables.
Solving took 0.0142 seconds.

Cost
----
 0.2561

Free Variables
--------------
x : 0.8443 
y : 0.8443 
z : 0.2561
```

So, why is the cost different and what are those sensitivity warnings about?

**Comment by bqpd** [comment]
what is Model.program in both cases?

**Comment by bqpd** [comment]
(answers his own question)

J as objective:

```python
gpkit.GeometricProgram(
  # minimize
    0.00329*x**-0.66*y**1.3 + 0.00329*x**1.3*y**-0.66 + 0.329*x**1.3*y**1.3 + 0.336*x**-0.66*y**-0.66,
[ # subject to
])
```

J as constraint:
```python
gpkit.GeometricProgram(
  # minimize
    z,
[ # subject to
    0.00329*x**-0.66*y**1.3*z**-0.33 + 0.00329*x**1.3*y**-0.66*z**-0.33 + 0.329*x**1.3*y**1.3*z**-0.33 + 0.336*x**-0.66*y**-0.66*z**-0.33 <= 1,
])
```

**Comment by bqpd** [comment]
These are clearly not equivalent programs.

**Comment by pgkirsch** [comment]
Ooops

**Comment by bqpd** [comment]
Wait why does this work with .solve()???

**Comment by pgkirsch** [comment]
Because it's the GP approximation?

**Comment by bqpd** [comment]
Lol yup

**Comment by bqpd** [comment]
Sorry about that!

**Comment by bqpd** [comment]
gps solved when J is a constraint:

```python
[gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.0098*x**-0.02*z**-0.98 + 0.0098*x**2*z**-0.98 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.0124*x**-0.17*z**-0.83 + 0.0124*x**1.8*z**-0.83 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.0311*x**-0.47*z**-0.53 + 0.0311*x**1.5*z**-0.53 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.0957*x**-0.73*z**-0.27 + 0.0957*x**1.3*z**-0.27 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.219*x**-0.88*z**-0.12 + 0.219*x**1.1*z**-0.12 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.344*x**-0.95*z**-0.047 + 0.344*x**1*z**-0.047 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.427*x**-0.98*z**-0.018 + 0.427*x**1*z**-0.018 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.468*x**-0.99*z**-0.0066 + 0.468*x**1*z**-0.0066 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.487*x**-1*z**-0.0024 + 0.487*x**1*z**-0.0024 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.495*x**-1*z**-0.0009 + 0.495*x**1*z**-0.0009 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.498*x**-1*z**-0.00033 + 0.498*x**1*z**-0.00033 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.499*x**-1*z**-0.00012 + 0.499*x**1*z**-0.00012 <= 1,
 ]), gpkit.GeometricProgram(
   # minimize
     z,
 [ # subject to
     0.5*x**-1*z**-4.5e-05 + 0.5*x**1*z**-4.5e-05 <= 1,
 ])]
```

when J is an objective:

```python
[gpkit.GeometricProgram(
   # minimize
     0.0098*x**-0.02 + 0.0098*x**2,
 [ # subject to
 ]), gpkit.GeometricProgram(
   # minimize
     0.00993*x**-0.002 + 0.00993*x**2,
 [ # subject to
 ]), gpkit.GeometricProgram(
   # minimize
     0.00997*x**-0.00063 + 0.00997*x**2,
 [ # subject to
 ]), gpkit.GeometricProgram(
   # minimize
     0.00998*x**-0.00035 + 0.00998*x**2,
 [ # subject to
 ])]
```

**Comment by pgkirsch** [comment]
Wait we are no longer discussing the same test case.

(This is my fault for introducing a new one halfway down the thread).

**Comment by bqpd** [comment]
Oops!

On @whoburg's point, I think putting a signomial objective in as a constraint is actually the only way to approximate it: I don't think we can support signomial objectives any other way.

**Comment by pgkirsch** [comment]
But it does fine when the coefficient is bigger than `0.01`..

**Comment by bqpd** [comment]
Yeah, but it's only an approximate approximation. =P

**Comment by pgkirsch** [comment]
So just to be clear, the proposed solution is to not permit signomial objectives

**Comment by bqpd** [comment]
Yeah. Does the non-objective version work for Woody?

**Comment by pgkirsch** [comment]
No :(

**Comment by pgkirsch** [comment]
Another case where the signomial-in-objective fails but the signomial-in-constraints succeeds.

```
import gpkit
from gpkit.shortcuts import *
x = Var('x')
with gpkit.EnableSignomials():
    m = Model(1-x, [x<=0.5])
    sol = m.localsolve()

# output

Cost
----
 0.6667

Free Variables
--------------
x : 0.5 
```

```
import gpkit
from gpkit.shortcuts import *
x = Var('x')
z = Var('z')
with gpkit.EnableSignomials():
    m = Model(z, [z +x >= 1, x<=0.5])
    sol = m.localsolve()

# output
Solving took 4 GP solves and 0.0777 seconds.

Cost
----
 0.5

Free Variables
--------------
x : 0.5 
z : 0.5 
```

**Comment by bqpd** [comment]
@whoburg, the error you got above of "GPs can't contain Signomials" seems like a `nan` or `inf` snuck into `x0` from the previous gp's result (see the `ravel()` warning)

**Comment by pgkirsch** [comment]
@bqpd With the re-factor it seems unnecessarily complicated to raise an exception that is specific to a problem type. For example, because of the `try/except` statements in `model.py`, I can't raise a `ValueError` in either `signomial_program._ _ init _ _()` or `signomial_program.localsolve()` without it escaping back to the `'localsolve()' can only be called on models that contain Signomials, because....` `ValueError`, which isn't what I want.

**Comment by bqpd** [comment]
Yeah it's gross :( I'll put in a better check, which PR do you want it in?

**Comment by pgkirsch** [comment]
I don't mind, is the test274 branch messy or would that work ok?

**Comment by bqpd** [comment]
That one might not be for merging, since it's got that expensive test. I'll put it in your PR, #337

**Comment by pgkirsch** [comment]
That works for me, thanks

**Comment by bqpd** [comment]
Okay, so I think I figured out one way to fix the bug @whoburg and the Jenkins machines are seeing, after reproducing on my windows laptop: as the objective approaches 0, the approximation (and the solvers) go a bit bonkers. cvxopt and mosek versions (or 32/64bitness) might explain why it works on some computers and not others? But it does seem like it either works for _all_ solvers on a comp or none.

The constraints don't get pushed to 0, but the objective does; I propose we stop SP solving at some eps > 0. For my windows laptop, `1e-6` is the smallest order of magnitude that works.

**Comment by bqpd** [comment]
Hmm, actually, `x - y <= 1` could get bad if another constraint is `1 + x <= 1`...

**Comment by whoburg** [comment]
Catching up. A few comments.

Re @pgkirsch's example, are you guys aware that the second program uses `gp = Model(z, [J <=1])`. Should that be  `gp = Model(z, [J <=z])` ?

Re: no signomials in objective, I agree this is the right policy for now.

**Comment by whoburg** [comment]
re: "Hmm, actually, `x - y <= 1` could get bad if another constraint is `1 + x <= 1`...we could stop if any posy_approx is below eps, not just the cost?", I don't follow.

**Comment by pgkirsch** [comment]
Wow whoops.. correct that and corrected the output

**Comment by bqpd** [comment]
@whoburg, I was basically just saying what you say elsewhere: if we want to allow signomials to approach 0, we'll have to do some extra work.

**Comment by bqpd** [comment]
Closing; we should open a new issue about 0 signomials
